content/blog/2019-11-17-chlorophyll/index.md

```diff
@@ -0,0 +1,236 @@
++++
+title = "Chlorophyll: Synthesis-Aided Compiler for Low-Power Spatial Architectures"
+extra.author = "Shaojie Xiang and Katy Voor"
+extra.bio = """
+  [Shaojie Xiang](https://github.com/Hecmay) is a 2nd year ECE PhD student researching on programming language and distributed system. 
+  [Katy Voor](https://github.com/kavoor) is a senior interested in compilers and computer architecture.
+"""
++++
+
+===
+## Introduction 
+
+Energy efficiency has been increasingly important for embedded processors and internet of things (IoT) systems. Many spatial hardware architectures have been proposed to address the power consumption problem without compromising the performance. GreenArrays 144 (GA144) is one example of the low power spatial processor. GA144 is a stack-based 18-bit processor without shared memory and fixed clock. To program GA144, users have to write an assembly-like low level programming language [[1]](#1) to define the data storage, data movement and communication logic to make use of the spatial architecture. This paper proposes a programming model and synthesis-aided compiler to compile high level C-like programs into binary executables on GA144. The compiler-optimized program can achieve comparable results with an expert written low level program.
+
+### Background 
+
+Spatial hardware is an architecture where the users are required to assign data, computations and communications explicitly to use its computing resources, storage and interconnect network. Spatial architecture is widely used in domains like High Performance Computing and image processing for its high performance and low power consumption. However, it is especially hard to program and debug due to low programmability and slow compile time. 
```

Comment:
I’m actually not sure I agree that spatial architectures are “widely used” in any domain yet. What sort of processors are you referring to? GreenArrays chips, for example, are not exactly common place. I think they probably *will* be, but unless I’m mistaken, we’re not there yet.

Reply:
It might be nice to briefly explain why this model leads to “low programmability” and “slow compile time.” What is “wrong” with the state of the art?

Reply:
Actually I am not so sure about the definition of spatial architecture, but in my understanding, FPGA is typical spatial architecture and is used as low power accelerators in many domains.

Reply:
Yeah, good question. It’s not a terribly well defined term, and classifying FPGAs as “just another” spatial architecture is somewhat controversial (although correct IMO). FPGAs are widespread but their use for actual *computation* (as opposed to circuit controls) is somewhat more niche.

Reply:
Sure. This paper is more focused on using spatial architecture as circuit controller. I guess I should rephrase a bit to avoid the potential misunderstanding on this term.